include/miopen/miopen.h

```diff
@@ -6582,6 +6582,77 @@ MIOPEN_EXPORT miopenStatus_t miopenBackendInitialize(miopenBackendDescriptor_t d
 // CLOSEOUT BackendAPI DOXYGEN GROUP
 #endif // MIOPEN_BETA_API
 
+#ifdef MIOPEN_BETA_API
+
+/** @addtogroup Cumulative Reduction
+ *
+ *  @{
+ */
+
+/*! @enum miopenCumOp_t
+ * Cumulative Reduction operation types
+ */
+using miopenCumOp_t = enum {
+    MIOPEN_CUM_MAX = 1,  /*!< the operation is getting the maximun value and index of the reduced
+                        elements */
+    MIOPEN_CUM_MIN = 2,  /*!< the operation is getting the minimum value and index of the reduced
+                        elements */
+    MIOPEN_CUM_SUM  = 3, /*!< the operation is getting the sum of the reduced elements */
+    MIOPEN_CUM_PROD = 4, /*!< the operation is getting the product of the reduced elements */
+};
+
+/*! @brief Helper function to query the minimum workspace size required by the Cumulative Reduction
+ * forward call
+ *
+ * @param handle                   MIOpen Handle (input)
+ * @param inputDesc                Tensor descriptor for input tensor (input)
+ * @param indicesDesc              Tensor descriptor for indices tensor (input)
+ * @param dim                      The dimension to do the operation over (input)
+ * @param sizeInBytes              Pointer to data to return the minimum workspace size
+ * @return                         miopenStatus_t
+ */
+MIOPEN_EXPORT miopenStatus_t
+miopenGetCumulativeReductionForwardWorkspaceSize(miopenHandle_t handle,
+                                                 miopenTensorDescriptor_t inputDesc,
+                                                 miopenTensorDescriptor_t indicesDesc,
+                                                 int dim,
+                                                 size_t* sizeInBytes);
+
+/*! @brief Execute a Cumulative Reduction forward layer
+ *
+ * @param handle                   MIOpen handle (input)
+ * @param workspace                Address of the allocated workspace data (input)
+ * @param workspaceSizeInBytes     Size in bytes of the allocated workspace data (input)
+ * @param inputDesc                Tensor descriptor for input tensor (input)
+ * @param input                    Data tensor input (input)
+ * @param outputDesc               Tensor descriptor for output tensor (input)
+ * @param output                   Data tensor output (output)
+ * @param indicesDesc              Tensor descriptor for indices tensor (input)
```

Comment:
It appears that in both solver and kernel, indices and output can be used optionally. 
If it's truly optional, please indicate it in the comments.
And could you please provide an example of how it works when the indices tensor is not used?

Additionally, I have allowed tensor inputs to be null when creating Adam. 
You don't have to follow it exactly, I will provide a link for reference.
https://github.com/moreh-dev/MIOpen/blob/develop/src/adam_api.cpp

Reply:
Well, I think indices and output should be prepared for being unused. Indices can be unused when binary operators are Sum or Product. Output can be unused with some operators similar to Argmin, Argmax. I haven't seen any operation like that yet but implementing for that case is very simple, so I just did it.

When the indices tensor is not used, users just have to use a `nullptr` as pointer to the indices tensor memory space. The HIP kernel then handles as indices tensor never exists. Other warpers such as miopen api work normally.

I appreciate your work with the Adam api. However, I think a unify api, function may be more flexible for user.